src/bucket/BucketManagerImpl.cpp

```diff
@@ -894,20 +894,20 @@ BucketManagerImpl::addBatch(Application& app, uint32_t currLedger,
 #ifdef BUILD_TESTS
     if (mUseFakeTestValuesForNextClose)
     {
-        currLedgerProtocol = mFakeTestProtocolVersion;
+        header.ledgerVersion = mFakeTestProtocolVersion;
     }
 #endif
     auto timer = mBucketAddBatch.TimeScope();
     mBucketObjectInsertBatch.Mark(initEntries.size() + liveEntries.size() +
                                   deadEntries.size());
-    mBucketList->addBatch(app, currLedger, currLedgerProtocol, initEntries,
-                          liveEntries, deadEntries);
+    mBucketList->addBatch(app, header.ledgerSeq, header.ledgerVersion,
+                          initEntries, liveEntries, deadEntries);
     mBucketListSizeCounter.set_count(mBucketList->getSize());
 
     if (app.getConfig().isUsingBucketListDB())
     {
         mSnapshotManager->updateCurrentSnapshot(
-            std::make_unique<BucketListSnapshot>(*mBucketList, currLedger));
+            std::make_unique<BucketListSnapshot>(*mBucketList, header));
```

Comment:
good call, better to avoid this footgun, so I moved this call to LedgerManager (called after advancing ledger pointers)

Reply:
I think we might still be updating the snapshot improperly in the catchup path. It looks like we call `updateCurrentSnapshot` at the end of assume state. This is fine for the `LedgerManagerImpl::loadLastKnownLedger` flow, but it looks like we end up snapshotting the same ledger twice in the catchup flow, where we snapshot once in assume state during the download-verify-apply step, then again when we call `setLastClosedLedger` on `CatchupWork::516`. I could be misreading the catchup state machine flow, but I don't think we should be updating snapshots in `assumeState`.

Reply:
Hm, so I think the contract of `updateCurrentSnapshot` changed recently. Before it was safe to update the snapshot of the same ledger, since it was just a simple swap of objects. After merging https://github.com/stellar/stellar-core/pull/4350, the new contract is that we must call `updateCurrentSnapshot` only once when the snapshot is fully constructed. That being said, I agree a cleanup is needed, so now there's a single call in `updateCurrentSnapshot`, that should cover all the cases.